### PR TITLE
Session4

### PR DESCRIPTION
**Pull request by valentinchaud**

### CODE REVIEW

File: Day2/README.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+Correct = 251.0
+Estimated = 263.84644
+********
+Correct = 158.0
+Estimated = 119.0271
+********
+Correct = 167.0
+Estimated = 179.02368
+
+Got that after final training :
```
Comment (bluesky0906) [inline Day2/README.md:80]: As we discussed in the session, I'm leaving this as a comment for a reminder:

When we monitor training progress, let's draw a learning curve.

File: src/Session3/LinearRegression.hs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+cost ::
+    Tensor -> -- ^ ground truth: 1 × 10
+    Tensor -> -- ^ estimated values: 1 × 10
+    Tensor    -- ^ loss: scalar
+cost z z' =  mean $ pow (2 :: Float)  (sub z z')
+
+calculateNewA :: Tensor -> Tensor -> Tensor -> Tensor -> Tensor
+calculateNewA x y slope intercept =
+    let predicatedValues = linear (slope, intercept) x
```
Comment (bluesky0906) [inline src/Session3/LinearRegression.hs:24]: Consider avoiding the same function with the same input in multiple locations. (L30)

File: src/Session4/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+## Part 1 : PerceptronAndGate
+
+For the perceptron and gate, as we can't modify values in Haskell due to how pure functions works, we have
+two solutions :
+
+1) We can use monads to modify the values
+2) We can use a recursive function that contains the history of weights and bias and add to the list the updated weights
+
+I decided to use the recursive function because it's still a bit difficult for me to use monads, and because it's 
+simpler when debugging because we can follow how the weights and bias are updated.
```
Comment (bluesky0906) [inline src/Session4/README.md:10]: I think we don't need to use monads for mutable values unless there's a specific reason:)

File: src/Session4/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+The first code uses MLPSpec for specifying the activation function and the number of neurons per layers. MLP contains
+the linear layers themselves. The MLP function do the forward pass part by applying the weights and bias, then the activation 
+function (here tanh). The main code generate random datas for the input of the xor gate, then apply mlp function to do the 
+forward propagation, before calculating the loss and using gradiant descent to the model's state (backpropagation part).
+It then updates weights and bias accordingly before going again until the model is optimized.
+
+The second code uses sigmoid instead of the tanh function. In the second code, we don't generate random datas for the input,
+and we calculate the loss in the entire trainingData instead, similarly to what I done for the loss calculation in my simple
+perceptron. It also uses CUDA, that will probably lead to better performance as it can use optimized operations for Nvidia GPUs.
```
Comment (bluesky0906) [inline src/Session4/README.md:43]: Do you have your own GPU machine!? It's great to know!

File: src/Session4/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+Sum of initial errors: Tensor Float []  3.0000
+Final weights: Tensor Float [1,2] [[ 2.0000   ,  4.0000   ]]
+Final bias: Tensor Float [1,1] [[-4.1000   ]]
+Final perceptronTensor Float [4,1] [[ 1.0000   ],
+                    [ 0.0000],
+                    [ 0.0000],
+                    [ 0.0000]]
+Final errorTensor Float []  0.0000
+```
+We can see that we go from 3 to 0 error value, so the code seems to work well after 60 epochs.
```
Comment (bluesky0906) [inline src/Session4/README.md:28]: It's good to add learning curve.

File: src/Session4/PerceptronAndGate.hs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+train::
+   Tensor -> -- weights
+   Tensor -> -- bias
+   Float ->  -- learningRate
+   Int ->    -- max epochs
+   Int ->    -- current epoch
+   [(Tensor, Tensor)] -> -- history of weights
+   [(Tensor, Tensor)]    -- new weights history
+train weights b learningRate maxEpochs currentEpoch history =
+  if currentEpoch >= maxEpochs
```
Comment (bluesky0906) [inline src/Session4/PerceptronAndGate.hs:35]: nit: you can use guards.

File: src/Session4/PerceptronAndGate.hs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        errors = calculateError expectedOutputs predictedOutputs
+
+        weightUpdates = mulScalar learningRate (matmul (transpose2D errors) inputs)
+
+        newWeights = add weights weightUpdates
+
+        biasUpdate = mulScalar learningRate (sumAll errors)
+        newBias = add b biasUpdate
+
+        newHistory = history ++ [(newWeights, newBias)]
```
Comment (bluesky0906) [inline src/Session4/PerceptronAndGate.hs:50]: nit: for the performance
```suggestion
        newHistory = (newWeights, newBias):history
```

File: src/Session4/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+The second code uses sigmoid instead of the tanh function. In the second code, we don't generate random datas for the input,
+and we calculate the loss in the entire trainingData instead, similarly to what I done for the loss calculation in my simple
+perceptron. It also uses CUDA, that will probably lead to better performance as it can use optimized operations for Nvidia GPUs.
+
+Using different activation function can lead to different results. For example, the step function won't work well because
+the gradient descent uses partial differentiation, but the derivate of 0 and 1 is 0, so the weights won't update because of that, and
+the learning will stop.
+
+Others functions will work. The main difference between sigmoid and tanh is that tanh will converge faster due to being centered
```
Comment (bluesky0906) [inline src/Session4/README.md:49]: Why do you think being centered to zero contributes to faster convergence?

**Review by bluesky0906 — COMMENTED** [review]
Hi,

As I mentioned in the seminar, please ensure your repository structure follows the instructions. (Please refer to the "Requests from Us" section in the following link: https://bekkilab.notion.site/2025-Introduction-to-Neural-Language-Processing-using-hasktorch-1be244489f42803d9496e813a7333843?pvs=4)

Overall, your work was truly excellent. Thank you for working on this despite your busy schedule!
